rpcs3/Emu/Cell/Modules/cellAtracXdec.cpp

```diff
@@ -0,0 +1,854 @@
+							{
+								output_s32[sample_idx * decoder.nch_in + ATXDEC_AVCODEC_CH_MAP[decoder.ch_config_idx - 1][channel_idx]] = static_cast<s32>(std::floor(sample * 0x80000000u));
+							}
+						}
+					}
+				}
+
+				first_decode = false;
+
+				if (error != CELL_OK)
+				{
+					// Block savestate creation during callbacks
+					std::unique_lock savestate_lock{g_fxo->get<hle_locks_t>(), std::try_to_lock};
+
+					if (!savestate_lock.owns_lock())
+					{
+						return;
+					}
+
+					skip_next_frame = true;
+					notify_error.cbFunc(ppu, error, notify_error.cbArg);
+				}
+			}
+
+			// Block savestate creation during callbacks
+			std::unique_lock savestate_lock{g_fxo->get<hle_locks_t>(), std::try_to_lock};
+
+			if (!savestate_lock.owns_lock())
+			{
+				return;
+			}
+
+			skip_getting_command = false;
+
+			// au_done and pcm_out callbacks are always called after a decode command, even if an error occurred
+			// The output always has to be consumed as well
+			notify_au_done.cbFunc(ppu, cmd.pcm_handle, notify_au_done.cbArg);
+
+			output_locked = true;
+			output_mutex_lock.unlock();
+
+			const u32 output_size = decoded_samples_num * (decoder.bw_pcm & 0x7fu) * decoder.nch_out;
+
+			const vm::var<CellAdecAtracXInfo> bsi_info{{ decoder.sampling_freq, decoder.ch_config_idx, decoder.nbytes }};
+
+			AdecCorrectPtsValueType correct_pts_type = ADEC_CORRECT_PTS_VALUE_TYPE_UNSPECIFIED;
+			switch (decoder.sampling_freq)
+			{
+			case 32000u: correct_pts_type = ADEC_CORRECT_PTS_VALUE_TYPE_ATRACX_32000Hz; break;
+			case 44100u: correct_pts_type = ADEC_CORRECT_PTS_VALUE_TYPE_ATRACX_44100Hz; break;
+			case 48000u: correct_pts_type = ADEC_CORRECT_PTS_VALUE_TYPE_ATRACX_48000Hz; break;
+			}
+
+			notify_pcm_out.cbFunc(ppu, cmd.pcm_handle, output, output_size, notify_pcm_out.cbArg, vm::make_var<vm::bcptr<void>>(bsi_info), correct_pts_type, error);
+			break;
+		}
+		default:
+			fmt::throw_exception("Invalid command");
+		}
+	}
+}
+
+template <AtracXdecCmdType type>
+error_code AtracXdecContext::send_command(ppu_thread& ppu, auto&&... args)
+{
+	ppu.state += cpu_flag::wait;
+
+	{
+		std::lock_guard lock{queue_mutex};
+
+		if constexpr (type == AtracXdecCmdType::close)
+		{
+			// Close command is only sent if the queue is empty on LLE
+			if (!cmd_queue.empty())
+			{
+				return {};
+			}
+		}
+
+		if (cmd_queue.full())
+		{
+			return CELL_ADEC_ERROR_ATX_BUSY;
+		}
+
+		cmd_queue.emplace(std::forward<AtracXdecCmdType>(type), std::forward<decltype(args)>(args)...);
+	}
+
+	queue_not_empty.notify_one();
+
+	return CELL_OK;
+}
+
+void atracXdecEntry(ppu_thread& ppu, vm::ptr<AtracXdecContext> atxdec)
+{
+	atxdec->decoder.alloc_avcodec();
+
+	atxdec->exec(ppu);
+
+	atxdec->decoder.free_avcodec();
+
+	if (thread_ctrl::state() == thread_state::aborting)
+	{
+		// For savestates, restore argument
+		idm::get<named_thread<ppu_thread>>(static_cast<u32>(atxdec->thread_id))->cmd_list
+		({
+			{ ppu_cmd::set_args, 1 }, static_cast<u64>(atxdec.addr()),
```

Comment:
1. Dont check thread_state::state for it, check cpu_flag::again state if it was raised inside exec(). (checking the reason the thread was aborted)
2. When ::again flag is set, whats saved in r3 to r6 is ppu_thread::syscall_args, so you rather save to syscall_args[0] instead of this.